src/krate.rs

```diff
@@ -638,68 +638,71 @@ impl Model for Crate {
-        let mut query = crates::table.select(ALL_COLUMNS).into_boxed();
-
-        if let Some(q_string) = params.get("q") {
-            let q = plainto_tsquery(q_string);
-            query = query.filter(q.matches(crates::textsearchable_index_col));
-        } else if let Some(letter) = params.get("letter") {
-            let pattern = format!("{}%", letter.chars().next().unwrap()
-                                           .to_lowercase().collect::<String>());
-            query = query.filter(canon_crate_name(crates::name).like(pattern));
-        } else if let Some(kw) = params.get("keyword") {
-            query = query.filter(crates::id.eq_any(
-                crates_keywords::table.select(crates_keywords::crate_id)
-                    .inner_join(keywords::table)
-                    .filter(lower(keywords::keyword).eq(lower(kw)))
-            ));
-        } else if let Some(cat) = params.get("category") {
-            query = query.filter(crates::id.eq_any(
-                crates_categories::table.select(crates_categories::crate_id)
-                    .inner_join(categories::table)
-                    .filter(categories::category.eq(cat).or(
-                            categories::category.like(format!("{}::%", cat))))
-            ));
-        } else if let Some(user_id) = params.get("user_id").and_then(|s| s.parse::<i32>().ok()) {
-            query = query.filter(crates::id.eq_any((
-                crate_owners::table.select(crate_owners::crate_id)
-                    .filter(crate_owners::owner_id.eq(user_id))
-                    .filter(crate_owners::owner_kind.eq(OwnerKind::User as i32))
-            )));
-        } else if params.get("following").is_some() {
-            query = query.filter(crates::id.eq_any((
-                follows::table.select(follows::crate_id)
-                    .filter(follows::user_id.eq(user?.id))
-            )));
-        }
-        Ok(query)
+    let mut query = crates::table
+        .select((ALL_COLUMNS, sql::<BigInt>("COUNT(*) OVER ()")))
+        .limit(limit)
+        .offset(offset)
+        .into_boxed();
+
+    if sort == "downloads" {
+        query = query.order(crates::downloads.desc())
+    } else {
+        query = query.order(crates::name.asc())
     }
 
-    let mut query = crates_query(&params, req.user())?;
-    let sort = params.get("sort").map(|s| &**s).unwrap_or("alpha");
-    match (params.get("q"), sort) {
-        (Some(q), "downloads") => {
-            query = query.order((crates::name.eq(q).desc(), crates::downloads.desc()))
-        }
-        (Some(q_string), _) => {
-            let q = plainto_tsquery(q_string);
+    if let Some(q_string) = params.get("q") {
+        let q = plainto_tsquery(q_string);
+        query = query.filter(q.matches(crates::textsearchable_index_col));
+
+        let perfect_match = crates::name.eq(q_string).desc();
+        if sort == "downloads" {
+            query = query.order((perfect_match, crates::downloads.desc()));
```

Comment:
Yes, calls to `order` override the earlier ones. It's actually unlike Active Record (`.reorder` replaces earlier ones, `.order` appends). The only method in Diesel which doesn't override earlier calls is `.filter`.